relay-system/src/service.rs

```diff
@@ -237,6 +243,316 @@ impl MessageResponse for NoResponse {
     }
 }
 
+/// Initial response to a [`BroadcastRequest`].
+#[derive(Debug)]
+enum InitialResponse<T> {
+    /// The response value is immediately ready.
+    ///
+    /// The sender did not attach to a broadcast channel and instead resolved the requested value
+    /// immediately. The request is now ready and can resolve. See [`BroadcastChannel::attach`].
+    Ready(T),
+    /// The sender is attached to a channel that needs to be polled.
+    Poll(Shared<oneshot::Receiver<T>>),
+}
+
+/// States of a [`BroadcastRequest`].
+enum BroadcastState<T> {
+    /// The request is waiting for an initial response.
+    Pending(oneshot::Receiver<InitialResponse<T>>),
+    /// The request is attached to a [`BroadcastChannel`].
+    Attached(Shared<oneshot::Receiver<T>>),
+}
+
+/// The request when sending an asynchronous message to a service.
+///
+/// This is returned from [`Addr::send`] when the message responds asynchronously through
+/// [`BroadcastResponse`]. It is a future that should be awaited. The message still runs to
+/// completion if this future is dropped.
+///
+/// # Panics
+///
+/// This future is not fused and panics if it is polled again after it has resolved.
+pub struct BroadcastRequest<T>(BroadcastState<T>)
```

Comment:
For my understanding: Could we merge `BroadcastRequest` and `BroadcastState` into a single enum here?

Reply:
We could, but then the internal states would become part of the public interface, and I wanted to prevent that.